### PR TITLE
Unify throwing of exceptions to use vw_exception instead of bare std::exception

### LINKED ISSUE #2168: VW should not throw exceptions without a message
We have a number of places in code which raise a std::exception with no message.

We should standardize on VW::vw_exception, and make sure we always have a message in the exception.

Also we should opportunistically get rid of `throw 1` in array_parameters.h

### PR DESCRIPTION
**Pull request by jackgerrits**
Fixes #2168

### CODE REVIEW

File: vowpalwabbit/array_parameters.h [hunk 1]
```diff
@@ -98,7 +99,10 @@ class sparse_parameters
   sparse_parameters(const sparse_parameters& other) { shallow_copy(other); }
   sparse_parameters(sparse_parameters&&) = delete;
 
-  weight* first() { throw 1; }  // TODO: Throw better exceptions. Allreduce currently not supported in sparse.
+  weight* first()
+  {
+    THROW_OR_RETURN_VOID("Allreduce currently not supported in sparse");
```
Comment (lokitoth) [inline vowpalwabbit/array_parameters.h]: This seems to be a weird macro name given that the function returns `weight*`. Is it the right one?

Comment (jackgerrits) [reply vowpalwabbit/array_parameters.h]: Oh yes good catch, thanks. This is wrong

File: vowpalwabbit/array_parameters.h [hunk 2]
```diff
@@ -101,7 +101,7 @@ class sparse_parameters
 
   weight* first()
   {
-    THROW_OR_RETURN_VOID("Allreduce currently not supported in sparse");
+    THROW_OR_RETURN_NORMAL("Allreduce currently not supported in sparse", nullptr);
```
Comment (lokitoth) [inline vowpalwabbit/array_parameters.h]: Is there a THROW_OR_RETURN() that functions differently you are avoiding colliding with?

Comment (jackgerrits) [reply vowpalwabbit/array_parameters.h]: Whoops, yes there is a macro that handles both these cases. Using that now.

**Pull request merged**